### PR TITLE
WIP: Refactor interpolation handling

### PR DESCRIPTION
**Pull request by mlshapiro**
See doc/source/developer/specs/interpolation.md for specification

### CODE REVIEW

**Comment by mlshapiro** [comment]
@jmilloy @mpu-creare please take a quick look at https://github.com/creare-com/podpac/blob/feature/interpolate/doc/source/developer/specs/interpolation.md#user-interface and provide feedback on usage interface. Note @mpu-creare I changed the terminology I was using for classes.  Please feel free to be critical of conventions or architecture

I am working on implementing `Interpolator` classes for what is already defined in DataSource and updating DataSource to use the `Interpolation` class

File: doc/source/developer/specs/data-source.md [hunk 1]
```diff
@@ -24,22 +24,39 @@
...
+- `_requested_source_coordinates_index` = tl.List()
+- `_requested_source_data` = tl.Instance(UnitsDataArray)
+
 #### Properties
 
 #### Contructor
 
 - FUTURE: After implementing a limiter on the request size, implement:
     + Take one input (i.e. `evaluate`) that will automatically execute the datasource at the native_coordinates on contruction. This will allow a shortcut when you just want to load a simple data source for processing with other more complication data sources
+- Choose a default interpolation option if neither interpolator or interpolation is defined
```
Comment (mpu-creare) [inline doc/source/developer/specs/data-source.md]: is `interpolator == _interpolator`?

File: doc/source/developer/specs/interpolation.md [hunk 2]
```diff
@@ -1,30 +1,208 @@
...
+    - Consider the cost of the interpolation operation
+- **TODO**: support custom interpolator classes?
+- **TODO**: support `+` (`__add__`) for Interpolators?
 
 # Example Use cases
-TODO: Add example usecases
 
-# Specification
+- user requests a single value at a point between coordinates in a datasource
+- user requests an array of coordinates from a dataset from a dataset with a different coordinate system
```
Comment (mpu-creare) [inline doc/source/developer/specs/interpolation.md]: Typo, 'from a dataset' 2x

File: doc/source/developer/specs/data-source.md (hunk 1 above)
Comment (mlshapiro) [reply doc/source/developer/specs/data-source.md]: this should read 'Choose a default interpolation option if `interpolation` is defined'

I have not pushed changes to DataSource spec - pushing now

**Comment by mlshapiro** [comment]
i just made a mess - i wont be done today

**Comment by mlshapiro** [comment]
Merged start into develop - continuing to work on feature/interpolators branch. 

@jmilloy please fix the TestPipiline and TestUnitDataArray tests on develop (or feel free to xfail the tests until they can be handled)

**Comment by mpu-creare** [comment]
Apparently I have to click an extra button for my comments to actually appear now. @mlshapiro  sorry of these are out of date.

**Comment by mlshapiro** [comment]
ah this is great thank you. I have ScipyGrid/Point working, i'm noticing that it does not work on the edge of the coordinate systems, looking into that now

**Pull request merged**